File: Lintcode 127 - Topological Sorting.java
Lintcode 127 - Topological Sorting.java

/**
 * Definition for Directed graph.
 * class DirectedGraphNode {
 *     int label;
 *     ArrayList<DirectedGraphNode> neighbors;
 *     DirectedGraphNode(int x) { label = x; neighbors = new ArrayList<DirectedGraphNode>(); }
 * };
 */
public class Solution {
    /**
     * @param graph: A list of Directed graph node
     * @return: Any topological order for the given graph.
     */    
    public ArrayList<DirectedGraphNode> topSort(ArrayList<DirectedGraphNode> graph) {
        // write your code here
        ArrayList<DirectedGraphNode> result = new ArrayList<DirectedGraphNode>();
        if (graph == null || graph.size() == 0) {
            return result;
        }
        
        //构建HashMap，记录所有的node及入度(即有几条边指向它)
        HashMap<DirectedGraphNode, Integer> map = new HashMap<DirectedGraphNode, Integer>();
        
        //这里注意，不是把所有的点加入HashMap，而是把它所有的neighbor加入
        for(DirectedGraphNode node: graph) {
            for(DirectedGraphNode nb : node.neighbors) {
                if(!map.containsKey(nb)) {
                    map.put(nb, 1);
                } else {
                    map.put(nb, map.get(nb) + 1);
                }
            }
        }
        
        Queue<DirectedGraphNode> Q = new LinkedList<DirectedGraphNode>();
        for(DirectedGraphNode node: graph) {
            //找到入度为0的点，第一个加入result
            if(!map.containsKey(node)) {
                result.add(node);
                Q.offer(node);
            }
        }
        
        while(!Q.isEmpty()) {
            DirectedGraphNode curr = Q.poll(); //这时，Q中所有的点的入度都是 0
            for(DirectedGraphNode nb : curr.neighbors) {
                map.put(nb, map.get(nb) - 1); //注意，所有的点得入度都需要 -1
                if(map.get(nb) == 0) {
                    Q.offer(nb);
                    result.add(nb);
                }
            }
        }
        
        return result;
    }
}

File: Lintcode 127 - Topological Sorting.cs
using System;
using System.Collections.Generic;

namespace CSharp
{
    class Node
    {
        public int label;
        public List<Node> neighbours;

        public Node(int label)
        {
            this.label = label;
            this.neighbours = new List<Node>();
        }

    }
    class Program
    {
        static List<Node> result = new List<Node>();
        static void Main(string[] args)
        {
            Node node0 = new Node(0);
            Node node1 = new Node(1);
            Node node2 = new Node(2);
            Node node3 = new Node(3);
            Node node4 = new Node(4);
            Node node5 = new Node(5);

            node0.neighbours = new List<Node>(){node1, node2, node3};
            node1.neighbours = new List<Node>(){node4};
            node2.neighbours = new List<Node>(){node4, node5};
            node3.neighbours = new List<Node>(){node4, node5};
            
            List<Node> graph = new List<Node>(){node0, node1, node2, node3, node4, node5};

            result = topSort(graph);

            foreach(Node node in result)
            {
                System.Console.Write(node.label + ", ");
            }
        }

        static List<Node> topSort(List<Node> graph)
        {
            Dictionary<Node, int> map = new Dictionary<Node, int>();
            Queue<Node> q = new Queue<Node>(0);

            //1. Build the map
            foreach(Node node in graph)
            {
                foreach(Node neighbour in node.neighbours)
                {
                    if(!map.ContainsKey(neighbour))
                    {
                        map[neighbour] = 1;
                    }
                    else
                    {
                        map[neighbour] = map[neighbour] + 1;
                    }
                }
            }

            foreach(Node node in graph)
            {
                if(!map.ContainsKey(node))
                {
                    result.Add(node);
                    q.Enqueue(node);
                }
            }

            while(q.Count > 0)
            {
                Node curr = q.Dequeue();
                foreach(Node node in curr.neighbours)
                {
                    map[node] = map[node] - 1;
                    if(map[node] == 0)
                    {
                        result.Add(node);
                        q.Enqueue(node);
                    }
                }
            }

            return result;
        }
    }
}
